core/arch/instr.h

```diff
@@ -498,6 +498,37 @@ struct _instr_t {
 };     /* instr_t */
 #endif /* DR_FAST_IR */
 
+/**
+ * A version of #instr_t which guarantees to not use heap allocation for regular
+ * decoding and encoding.  It inlines all the possible operands and encoding space
+ * inside the structure.  Some operations could still use heap if custom label data is
+ * used to point at heap-allocated structures through extension libraries or custom
+ * code.
+ *
+ * The 'instr' field should be passed to decoding and other routines.  E.g.:
+ *
+ *    instr_noalloc_t noalloc;
+ *    instr_noalloc_init(dcontext, &noalloc);
+ *    pc = decode(dcontext, ptr, &noalloc.instr);
```

Comment:
Do we instead want to make instr_noalloc_t opaque from the user's perspective and create a new API function to obtain the instr field?

Reply:
Yes that does seem nicer.

Reply:
Done.